### PR TITLE
drivers: led_strip: Add support for external SPI CS on APA102 LED strips

### PR DESCRIPTION
**Pull request by NZSmartie**
Some boards may include an APA102 LED on an existing SPI bus. With
additional circuitry, chip select is supported where the APA102
lacks a built in CS pin.

> image
Example APA102 LED with external chip select by utilizing a dual non-inverting line buffer (74LVC2G125)

This patch allows CS to be defined in the devicetree (through .dts
files) and utilised in the apa102.c driver if it's available. Preserving
old behaviour if no chip select is defined.

Signed-off-by: Roman Vaughan <nzsmartie@gmail.com>

### CODE REVIEW

File: drivers/led_strip/apa102.c [hunk 1]
```diff
@@ -93,6 +96,23 @@ static int apa102_init(struct device *dev)
...
+		device_get_binding(DT_INST_SPI_DEV_CS_GPIOS_LABEL(0));
+	if (!data->cs_ctl.gpio_dev) {
+		return -ENODEV;
+	}
+	data->cs_ctl.gpio_pin = DT_INST_SPI_DEV_CS_GPIOS_PIN(0);
+	data->cs_ctl.delay = 0;
+
+	data->cfg.cs = &data->cs_ctl;
+
+	int flags = DT_INST_SPI_DEV_CS_GPIOS_FLAGS(0);
```
Comment (mbolivar-nordic) [inline drivers/led_strip/apa102.c:110]: Although it's rare, not all GPIO specifiers support a flags field. Can you please add the appropriate ifdeffery to support this case? See samples/basic/blinky/src/main.c and samples/basic/button/src/main.c for two different ways to achieve this.

**Review by mbolivar-nordic — COMMENTED** [review]
Thanks for the patch! I hadn't seen this trick before. Clever.

I don't have an APA102 to test on, so can you confirm you've tested this on hardware with and without CS set up in the DTS?

If so, other than the flags change, I am +1 to get this merged when the 2.4 cycle opens up the master branch agian.

File: drivers/led_strip/apa102.c (hunk 1 above)
Comment (NZSmartie) [reply drivers/led_strip/apa102.c:110]: At the time of writing, there isn't a clear way to check if the flags field exists without making the code some-what unreadable. Instead #25827 would be a good way to address this

File: drivers/led_strip/apa102.c [hunk 2]
```diff
@@ -9,10 +9,31 @@
...
+
+/*
+ * Devicetree helper macros which gets the 'flags' cell from a 'cs_gpios'
+ * property on the spi bus, or returns 0 if the property has no 'flags' cell.
+ */
+#define HAS_FLAGS(spi_dev, spi_reg)					\
+	DT_PHA_HAS_CELL_AT_IDX(spi_dev, cs_gpios, spi_reg, flags)
+
+#define DT_INST_SPI_DEV_CS_GPIOS_HAS_FLAGS(node)			\
+	HAS_FLAGS(DT_BUS(node), DT_REG_ADDR(node))
```
[the remark this answers is not in the document]
Comment (NZSmartie) [reply drivers/led_strip/apa102.c]: Makes sense, reserved prefix has been removed

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by mbolivar-nordic** [comment]
Looks ready to merge; I'm assuming the buildkite failure can be ignored

**Comment by NZSmartie** [comment]
Have rebased with master, buildkite should work now.

**Comment by mbolivar-nordic** [comment]
@Mani-Sadhasivam I know this is led_strip and not led, but I'm hoping you can give this a review as well. Now that we need two +1's to merge a PR, this is blocked in review.

File: drivers/led_strip/apa102.c [hunk 3]
```diff
@@ -93,6 +116,21 @@ static int apa102_init(struct device *dev)
 	data->cfg.operation =
 		SPI_OP_MODE_MASTER | SPI_TRANSFER_MSB | SPI_WORD_SET(8);
 
+#if DT_INST_SPI_DEV_HAS_CS_GPIOS(0)
+	data->cs_ctl.gpio_dev =
+		device_get_binding(DT_INST_SPI_DEV_CS_GPIOS_LABEL(0));
+	if (!data->cs_ctl.gpio_dev) {
+		return -ENODEV;
+	}
```
Comment (Mani-Sadhasivam) [inline drivers/led_strip/apa102.c:124]: Just a style nit: I'd add a new line after `}`

**Review by Mani-Sadhasivam — APPROVED** [review]
LGTM

**Comment by mbolivar-nordic** [comment]
Thanks @Mani-Sadhasivam !

**Pull request merged**